motionplan/motionPlanner.go

```diff
@@ -385,3 +382,90 @@ IK:
 	}
 	return orderedSolutions, nil
 }
+
+// CheckPlan checks if obstacles intersect the trajectory of the frame following the plan.
+func CheckPlan(
+	checkFrame frame.Frame,
+	plan []map[string][]frame.Input,
+	worldState *frame.WorldState,
+	fs frame.FrameSystem,
+	errorState spatialmath.Pose,
+) error {
+	// since there are only two states we care about, lets just return err
+	// ensure that we can actually perform the check
+	if len(plan) < 2 {
+		return errors.New("plan must have at least two elements")
+	}
+
+	// construct solverFrame
+	// Note that this requires all frames which move as part of the plan, to have an
+	// entry in the very first plan waypoint
+	sf, err := newSolverFrame(fs, checkFrame.Name(), frame.World, plan[0])
+	if err != nil {
+		return err
+	}
+
+	// construct planager
+	sfPlanner, err := newPlanManager(sf, fs, golog.NewLogger("checkPlan-logger"), defaultRandomSeed)
+	if err != nil {
+		return err
+	}
+
+	// convert plan into nodes
+	planNodes := make([]node, 0, len(plan))
+	for _, step := range plan {
+		stepConfig, err := sf.mapToSlice(step)
+		if err != nil {
+			return err
+		}
+		pose, err := sf.Transform(stepConfig)
+		// adjust pose based off how much we've deviated from the expected path
+		pose = spatialmath.Compose(pose, errorState)
+		if err != nil {
+			return err
+		}
+		planNodes = append(planNodes, &basicNode{q: stepConfig, pose: pose})
+	}
+
+	// This should be done for any plan whose configurations are specified in relative terms rather than absolute ones.
+	// Currently this is only TP-space, so we check if the PTG length is >0.
+	// The solver frame will have had its PTGs filled in the newPlanManager() call, if applicable.
+	relative := len(sf.PTGs()) > 0
+
+	if relative {
+		if planNodes, err = rectifyTPspacePath(planNodes, sf); err != nil {
+			return err
+		}
+	}
+	startPose := planNodes[0].Pose()
+	goalPose := planNodes[len(planNodes)-1].Pose()
+
+	// create constraints
+	if sfPlanner.planOpts, err = sfPlanner.plannerSetupFromMoveRequest(
+		startPose,
+		goalPose,
+		plan[0],
+		worldState,
+		nil, // no pb.Constraints
+		nil, // no plannOpts
+	); err != nil {
+		return err
+	}
+
+	// go through plan and check that we can move from plan[i] to plan[i+1]
+	for i := 0; i < len(planNodes)-1; i++ {
+		currentPose := planNodes[i].Pose()
```

Comment:
I purposefully chose to omit giving a `startconfiguration` and `endconfiguration` value to the `ik.Segment`.

this is because when our `errorState` is not `r3.Vector{0,0,0}` we cannot rely on information given to us through values of type `referenceframe.Input`

this decision, in turn, corresponds to the way I've set up the conditionals within `CheckStateConstraintsAcrossSegment`

Reply:
hmmm, ok.
Then it is unclear to me how TPspace plans can be checked when there an errorState which is not the zero pose.
My reasoning is that errorState allows us to only augment the pose, and not the inputs of a respective element of the plan.

Reply:
spoke offline with ray about this. I will see if there is a clean way for me to introduce a new type of constraint.. or do some math on the already existing constraint.